lib/api/completeMultipartUpload.js

```diff
@@ -153,235 +156,97 @@ function completeMultipartUpload(authInfo, request, log, callback) {
                 });
             }
-            return next(null, destBucket, objMD, mpuBucket, jsonList,
-            storedMetadata, null, mpuOverviewKey);
+            return next(null, destBucket, objMD, mpuBucket, storedParts,
+                jsonList, storedMetadata, null, mpuOverviewKey);
         },
-        function retrieveParts(destBucket, objMD, mpuBucket, jsonList,
-        storedMetadata, completeObjData, mpuOverviewKey, next) {
-            // means MPU completion was done on third-party backend,
-            // so part info is not in our metadata and related steps
-            // should be skipped
+        function validateAndFilterParts(destBucket, objMD, mpuBucket,
+        storedParts, jsonList, storedMetadata, completeObjData, mpuOverviewKey,
+        next) {
             if (completeObjData) {
-                return next(null, destBucket, objMD, mpuBucket, null,
-                jsonList, storedMetadata, completeObjData, mpuOverviewKey);
+                return next(null, destBucket, objMD, mpuBucket, storedParts,
+                jsonList, storedMetadata, completeObjData, mpuOverviewKey,
+                completeObjData.filteredPartsObj);
             }
-            return services.getMPUparts(mpuBucket.getName(), uploadId, log,
-                (err, result) => {
-                    if (err) {
-                        return next(err, destBucket);
-                    }
-                    const storedParts = result.Contents;
-                    return next(null, destBucket, objMD, mpuBucket,
-                        storedParts, jsonList, storedMetadata, null,
-                        mpuOverviewKey);
-                });
+            const filteredPartsObj = validateAndFilterMpuParts(storedParts,
+                jsonList, mpuOverviewKey, splitter, log);
+            if (filteredPartsObj.error) {
+                return next(filteredPartsObj.error, destBucket);
+            }
+            return next(null, destBucket, objMD, mpuBucket, storedParts,
+                jsonList, storedMetadata, completeObjData, mpuOverviewKey,
+                filteredPartsObj);
         },
         function processParts(destBucket, objMD, mpuBucket, storedParts,
-            jsonList, storedMetadata, completeObjData, mpuOverviewKey, next) {
-            let calculatedSize = 0;
-            if (completeObjData) {
+        jsonList, storedMetadata, completeObjData, mpuOverviewKey,
+        filteredPartsObj, next) {
+            // if mpu was completed on backend that stored mpu MD externally,
+            // skip MD processing steps
+            if (completeObjData && skipMpuPartProcessing(completeObjData)) {
                 const dataLocations = [
                     {
                         key: completeObjData.key,
                         size: completeObjData.contentLength,
-                        start: calculatedSize,
+                        start: 0,
                         dataStoreName: storedMetadata.dataStoreName,
```

Comment:
you could have add `dataStoreType` key  to be consistent with the locations schema.